In [2]:
%load_ext autoreload 
%autoreload 2

In [7]:
import pandas as pd 
#import geopandas as gpd 
import os 
import requests
import urllib
import io
import numpy as np 

In [2]:
os.getcwd()

'/Users/max/Documents/dcp-projects/db-equitable-development-tool/notebooks'

In [3]:
import sys 
sys.path.append('../utils')
import wd_management 
wd_management.set_wd_root()

In [4]:
hny_url = "https://data.cityofnewyork.us/api/views/hg8x-zxpr/rows.csv?accessType=DOWNLOAD"

hny = requests.get(hny_url).content

hny_df = pd.read_csv(io.StringIO(hny.decode('utf-8')))



In [5]:
hny_df.shape

(5362, 41)

#### Lets do some exploratory analysis to understand the data 
* We are particularily interested in the amount of records with an address but no lat long data in the respective columns - how much work will we need to do to geocode these records on our end?



In [8]:
# number and street are encoded with ---- insted of nan
hny_df = hny_df.replace('----', np.nan)

In [11]:
# how many records have any NANA 
#hny_df_na = hny_df[hny_df.isna().any(axis=1)]

#hny_df_na.shape


(3557, 41)

In [9]:
def show_missing(df):
    """
    Return the total missing values and the percentage of
    missing values by column. 
    """
    null_count = df.isnull().sum()
    null_percentage = (null_count / df.shape[0]) * 100
    return pd.DataFrame({'count_missing_records': null_count, 'missing_percentage': null_percentage})

In [10]:
show_missing(hny_df)

count_missing_records  missing_percentage
Project ID                                              0            0.000000
Project Name                                            0            0.000000
Project Start Date                                      0            0.000000
Project Completion Date                              1608           29.988810
Building ID                                          1021           19.041402
Number                                               1021           19.041402
Street                                               1021           19.041402
Borough                                                 0            0.000000
Postcode                                             1083           20.197687
BBL                                                  1154           21.521820
BIN                                                  2584           48.190974
Community Board                                         0            0.000000
Council District                                        8            0.149198
Census Tract                                         1077           20.085789
NTA - Neighborhood Tabulation Area                   1077           20.085789
Latitude                                             1083           20.197687
Longitude                                            1083           20.197687
Latitude (Internal)                                  1161           21.652369
Longitude (Internal)                                 1161           21.652369
Building Completion Date                             1392           25.960463
Reporting Construction Type                             0            0.000000
Extended Affordability Only                             0            0.000000
Prevailing Wage Status                                  0            0.000000
Extremely Low Income Units                              0            0.000000
Very Low Income Units                                   0            0.000000
Low Income Units                                        0            0.000000
Moderate Income Units                                   0            0.000000
Middle Income Units                                     0            0.000000
Other Income Units                                      0            0.000000
Studio Units                                            0            0.000000
1-BR Units                                              0            0.000000
2-BR Units                                              0            0.000000
3-BR Units                                              0            0.000000
4-BR Units                                              0            0.000000
5-BR Units                                              0            0.000000
6-BR+ Units                                             0            0.000000
Unknown-BR Units                                        0            0.000000
Counted Rental Units                                    0            0.000000
Counted Homeownership Units                             0            0.000000
All Counted Units                                       0            0.000000
Total Units                                             0            0.000000

In [11]:
hny_missing_geo = hny_df.loc[(hny_df['Latitude'].isnull()) & (hny_df['Street'].notnull())]

In [12]:
hny_missing_geo.shape

(62, 41)

In [13]:
from IPython.display import display 
pd.options.display.max_columns = None

In [14]:
display(hny_missing_geo)

Project ID                                       Project Name  \
3          44223                           ROCHESTER SUYDAM PHASE 1   
160        70806                              1718 PRESIDENT STREET   
194        69140                FORDHAM BEDFORD.FB CLUSTER.NR REHAB   
206        69285           3800 PUTNAM AVENUE GARDENS HDFC.HPO.FY21   
257        69331                                   3-11 27TH AVENUE   
...          ...                                                ...   
3677       53017  Nehemiah Spring Creek Homes at Gateway Estates 4A   
3678       53017  Nehemiah Spring Creek Homes at Gateway Estates 4A   
3679       53017  Nehemiah Spring Creek Homes at Gateway Estates 4A   
3944       49635            Habitat-NYC Single Family Homes Phase I   
4236       52722                                Our Lady of Lourdes   

     Project Start Date Project Completion Date  Building ID   Number  \
3            06/30/2021                     NaN     977564.0  329/331   
160          06/29/2021                     NaN     998916.0     1781   
194          06/25/2021                     NaN     122133.0       17   
206          06/23/2021              06/23/2021     863998.0     3800   
257          05/14/2021                     NaN     993634.0   11-Mar   
...                 ...                     ...          ...      ...   
3677         06/27/2016              05/28/2019     972193.0      503   
3678         06/27/2016              05/28/2019     972194.0      505   
3679         06/27/2016              05/28/2019     972195.0      507   
3944         12/31/2015              05/31/2016     607523.0   137-35   
4236         08/18/2015              06/29/2018     958049.0   21-Nov   

                          Street   Borough  Postcode           BBL  BIN  \
3                   RALPH AVENUE  Brooklyn       NaN           NaN  NaN   
160             PRESIDENT STREET  Brooklyn       NaN           NaN  NaN   
194   WEST MOSHOLU PARKWAY NORTH     Bronx       NaN           NaN  NaN   
206           PUTNAM AVENUE WEST     Bronx       NaN           NaN  NaN   
257                    27 AVENUE    Queens       NaN           NaN  NaN   
...                          ...       ...       ...           ...  ...   
3677           SCHROEDERS AVENUE  Brooklyn       NaN  3.044520e+09  NaN   
3678           SCHROEDERS AVENUE  Brooklyn       NaN  3.044520e+09  NaN   
3679           SCHROEDERS AVENUE  Brooklyn       NaN  3.044520e+09  NaN   
3944                   220 PLACE    Queens       NaN           NaN  NaN   
4236               DESALES PLACE  Brooklyn       NaN           NaN  NaN   

     Community Board  Council District Census Tract  \
3              BK-03              41.0          NaN   
160            BK-09              41.0          NaN   
194            BX-07              11.0          NaN   
206            BX-08              11.0          NaN   
257            QN-01              22.0          NaN   
...              ...               ...          ...   
3677           BK-05              42.0          NaN   
3678           BK-05              42.0          NaN   
3679           BK-05              42.0          NaN   
3944           QN-13              31.0          NaN   
4236           BK-04              37.0          NaN   

     NTA - Neighborhood Tabulation Area  Latitude  Longitude  \
3                                   NaN       NaN        NaN   
160                                 NaN       NaN        NaN   
194                                 NaN       NaN        NaN   
206                                 NaN       NaN        NaN   
257                                 NaN       NaN        NaN   
...                                 ...       ...        ...   
3677                                NaN       NaN        NaN   
3678                                NaN       NaN        NaN   
3679                                NaN       NaN        NaN   
3944                                NaN       NaN        NaN   
4236      

#### Conclusion 

There are only 62 records out of 5362 that have a missing lat/long but have some sort of street address. From a quick run through of these records, there look to be some encoding errors for certain records (house #'s becing encoded as dates i.e 3-11 27TH AVENUE becoming -->	11-Mar	27 AVENUE), some may also be from simple spelling errors that HPD geocoding wasn't able to catch. Can write this list and send back to HPD to rectify errors - Amanda doesn't think this will be too big of an issue. 

Although over 20% of records have missing lat/long and/or address level information, HPD provides us with the CD which can  be roughly mapped to PUMA level geography which for our purposes should be sufficient for the EDTT project.

